internal/gps/source_cache.go

```diff
@@ -128,11 +153,41 @@ func (c *singleSourceCacheMemory) setPackageTree(r Revision, ptree pkgtree.Packa
 	c.mut.Unlock()
 }
 
-func (c *singleSourceCacheMemory) getPackageTree(r Revision) (pkgtree.PackageTree, bool) {
+func (c *singleSourceCacheMemory) getPackageTree(r Revision, pr ProjectRoot) (pkgtree.PackageTree, bool) {
 	c.mut.Lock()
-	ptree, has := c.ptrees[r]
+	rptree, has := c.ptrees[r]
 	c.mut.Unlock()
-	return ptree, has
+
+	if !has {
+		return pkgtree.PackageTree{}, false
+	}
+
+	ptree := pkgtree.PackageTree{
+		ImportRoot: string(pr),
+		Packages:   make(map[string]pkgtree.PackageOrErr),
+	}
+	// Return a copy, with full import paths.
```

Comment:
Use `PackageTree.Copy()` here instead of re-rolling what's already there - even if it means another iteration pass over the packages.

Or, implement a new `CopyWithRoot()` func that does the same as here, then reimplement `Copy()` in terms of that.